core/src/block.rs

```diff
@@ -19,894 +19,635 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
-            #[allow(clippy::expect_used)]
-            let timestamp = crate::current_time().as_millis();
-            // TODO: Need to check if the `transactions` vector is empty. It shouldn't be allowed.
-            Self {
-                timestamp,
-                transactions,
-                event_recommendations,
-            }
-        }
+impl PendingBlockStub<'_> {
+    /// Create a new [`PendingBlock`] from transactions.
+    pub fn build(self) -> PendingBlock {
+        let timestamp = crate::current_time().as_millis();
+        // TODO: Need to check if the `transactions` vector is empty. It shouldn't be allowed.
+
+        let mut header = BlockHeader {
+            timestamp,
+            consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
+            height: self.height,
+            view_change_index: self.view_change_index,
+            previous_block_hash: self.previous_block_hash,
+            transactions_hash: None,
+            rejected_transactions_hash: None,
+            committed_with_topology: self.committed_with_topology.sorted_peers,
+        };
 
-        /// Chain block with the existing blockchain.
-        pub fn chain(
-            self,
-            height: u64,
-            previous_block_hash: Option<HashOf<VersionedCommittedBlock>>,
-            view_change_index: u64,
-            committed_with_topology: Topology,
-        ) -> ChainedBlock {
-            ChainedBlock {
-                transactions: self.transactions,
-                event_recommendations: self.event_recommendations,
-                header: BlockHeader {
-                    timestamp: self.timestamp,
-                    consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
-                    height: height + 1,
-                    view_change_index,
-                    previous_block_hash,
-                    transactions_hash: None,
-                    rejected_transactions_hash: None,
-                    committed_with_topology: committed_with_topology.sorted_peers,
-                },
-            }
-        }
+        let mut txs = Vec::new();
+        let mut rejected = Vec::new();
 
-        /// Create a new blockchain with current block as a first block.
-        pub fn chain_first_with_topology(self, genesis_topology: Topology) -> ChainedBlock {
-            ChainedBlock {
-                transactions: self.transactions,
-                event_recommendations: self.event_recommendations,
-                header: BlockHeader {
-                    timestamp: self.timestamp,
-                    consensus_estimation: DEFAULT_CONSENSUS_ESTIMATION_MS,
-                    height: 1,
-                    view_change_index: 0,
-                    previous_block_hash: None,
-                    transactions_hash: None,
-                    rejected_transactions_hash: None,
-                    committed_with_topology: genesis_topology.sorted_peers,
-                },
+        let wsv = self.wsv.clone();
```

Comment:
Pasting my comment from other conversation:
The cloning was introduce by someone other than me in order to solve a bug. This is a immutable reference but it infact has interior mutability. If we don't clone then the wsv is modified and that is incorrect behaviour. We need a local copy we can mutate because we need to validate a series of transactions and how they affect each other.

Reply:
The plan is to replace this with change sets. But that's will wait on general WSV refactoring.

Reply:
Gotcha, just was a bit confusing without some clarifying comment there. I do vaguely remember such conversation happening, though.

Reply:
Imo we can avoid such confusion in future by accepting `WorldStateView` as parameter, not `&WorldStateView`.

Reply:
Rust in general prefers allocations be pushed as far up the call stack as possible. Thus accepting a `wsv` by reference to then clone is considered a bad practice (with caveats).

Reply:
Why this was resolved? there is neither changes nor ticket or reasons why this won't work?
Imo this change is pretty small to be separate ticket.

Reply:
It felt like bikeshedding to me since it will be returned to a pass by reference anyway after we have implemented the new-wsv.

Reply:
whether or not it's bikeshedding, you should not be resolving discussions as you see fit. Discussion can only be closed if given permission by the OP or other. Mutual agreement has to be reached

Reply:
This change has been made now.